### PR TITLE
Raise user handler exception in MultiThreadedExecutor.

### PR DESCRIPTION
**Pull request by fujitatomoya**
addresses https://github.com/ros2/rclpy/issues/983

Signed-off-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

### CODE REVIEW

File: rclpy/rclpy/executors.py
Comment (fujitatomoya) [inline rclpy/rclpy/executors.py:765]: create object to raise the exception if there is, instead of managing the `future.result()` via `submit`.

**Comment by fujitatomoya** [comment]
@sloretz @ivanpauno could you take a look?

File: rclpy/rclpy/executors.py
Comment (ivanpauno) [inline rclpy/rclpy/executors.py:768]: I'm not sure if it's a great idea to throw an exception from the executor when any of the callbacks fail.
It would make everything fail when only one callback failed.

Maybe something else like logging an error is a good idea, I'm not sure.

Comment (mr337) [reply rclpy/rclpy/executors.py:768]: That does bring up a good point. The main issue I had in https://github.com/ros2/rclpy/issues/983 is silently failing, and the message that no retrieval of exception could be very delayed leading to head scratching.

If we decide to not to throw exception from the executor, would logging it out be better, assuming that would be immediate? As a preventative measure all the timer call backs in our code are wrapped in a `try except` block but was a odd experience figuring that out from a user perspective.

Comment (ivanpauno) [reply rclpy/rclpy/executors.py:768]: > If we decide to not to throw exception from the executor, would logging it out be better, assuming that would be immediate? 

Yes, logging is better than nothing.
And if you log from here that should be immediate.

Comment (sloretz) [reply rclpy/rclpy/executors.py:768]: +1 to logging. Logging and ignoring it has some precedent [in concurrent.futures](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.add_done_callback).

An enhancement might be to allow the user to set a custom exception handler, as [`asyncio` seems to allow people to do](https://docs.python.org/3/library/asyncio-eventloop.html#error-handling-api).

Comment (fujitatomoya) [reply rclpy/rclpy/executors.py]: addressed in https://github.com/ros2/rclpy/pull/984/commits/cda6aeeccb78777101235fafcd4ca6a14acc70d9

File: rclpy/rclpy/executors.py
Comment (fujitatomoya) [inline rclpy/rclpy/executors.py:729]: I think we should provide the same experience for `SingleThreadedExecutor`.

**Comment by fujitatomoya** [comment]
the followings are the output with https://github.com/ros2/rclpy/issues/983

- SingleThreadedExecutor

```bash
# ros2 run prover_rclpy rclpy_983
[WARN] [1660060486.243490700] [SingleThreadedExecutor]: Exception: division by zero
[WARN] [1660060487.218409974] [SingleThreadedExecutor]: Exception: division by zero
[WARN] [1660060488.218467063] [SingleThreadedExecutor]: Exception: division by zero
[WARN] [1660060489.218468133] [SingleThreadedExecutor]: Exception: division by zero
[WARN] [1660060490.218465565] [SingleThreadedExecutor]: Exception: division by zero
...<snip>
```

- MultiThreadedExecutor

```bash
# ros2 run prover_rclpy rclpy_983
[WARN] [1660060398.085604463] [MultiThreadedExecutor]: Exception: division by zero
[WARN] [1660060399.065598330] [MultiThreadedExecutor]: Exception: division by zero
[WARN] [1660060400.065630959] [MultiThreadedExecutor]: Exception: division by zero
[WARN] [1660060401.065543777] [MultiThreadedExecutor]: Exception: division by zero
[WARN] [1660060402.065565788] [MultiThreadedExecutor]: Exception: division by zero
<...snip>
```

File: rclpy/rclpy/executors.py
Comment (ivanpauno) [reply rclpy/rclpy/executors.py:729]: makes sense to me

**Review by ivanpauno — APPROVED** [review]
LGTM!

Thanks for the contribution

**Comment by iuhilnehc-ynos** [comment]
I don't like this fix that only uses a warning log. What if users want to exit if any callback got an exception?
It's totally the users' responsibility to catch the exceptions either in the callback or calling spin.

**Comment by fujitatomoya** [comment]
@ivanpauno @sloretz 

i am just concerned that what if user application expects the exception from user callback? there would be the case that user application intentionally generates the exception and catch that in user code. with this fix, it cannot be supported.
yet, i think this PR works for some cases, to keep the executors running with warnings.

just want to hear opinion about that before we push this fix further.

CC: @clalancette @wjwwood

**Comment by ivanpauno** [comment]
> i am just concerned that what if user application expects the exception from user callback? there would be the case that user application intentionally generates the exception and catch that in user code. with this fix, it cannot be supported.

I would say it's uncommon to throw an exception from a callback and catch it from `Executor.spin()`.
It's not a pattern that I have seen and I wouldn't encourage it, but it was possible for the `SingleThreadedExecutor`.

So it's true that in that case it can break pre-existing code.
If we do the opposite (throwing an exception by default), we could break the `MultiThreadedExecutor`, which by default did nothing.

I think consistency is important, so I would do the same in both.
I don't mind the default, but logging may be fine (what does rclcpp do?).
Having a customizable exception handler like @sloretz sounds like a good idea as well (which by default, would only log).

**Comment by fujitatomoya** [comment]
> what does rclcpp do?

good question, it just raises exception from user callback. (both `SingleThreadedExecutor` and `MultiThreadedExecutor`)

> I would say it's uncommon to throw an exception from a callback and catch it from Executor.spin().

agree.

i think we need to do,
- run CI for this PR. (will do it right now.)
- apply the same patch for `rclcpp` to log the exception message once executor
- consider more enhancement like customizable exception handler

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17113/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11661/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17530/)

**Comment by wjwwood** [comment]
In my opinion, the executors (applies to both c++ and Python), exceptions should not be caught by default. If there is an exception that the user did not catch themselves, it should go to the unhandled exception handler (which aborts the application). This matches what happens in plain python and c++. 

I think that assuming that "when an exception occurs in a callback, it is correct to catch it and continue execution" is not a good assumption and isn't safe either. 

Instead we should allow users to set their own executor level unhandled exception handler. This would be called when an exception is thrown in the callbacks called by the executor.

For situations like component mode containers we could install a custom unhandled exception handler that logs the exception and then unloads the node. This avoids taking down the whole application when an unhandled exception occurs in a single node. 

There can also be exceptions inside the executor that aren't part of the callback being called necessarily. But I think those cases can be pushed into two buckets: things that should be caught by wrapping the spin call, and providing callbacks in the entities when errors specific to them occur, e.g. a timeout occurs when waiting for a service response times out. 

So I wouldn't do this pr as is. I think both the multi-threaded and single-threaded executors should exit when an exception occurs in the user callback. 

I'm not sure if we should reroute the exception to the thread that called spin in the multi threaded case. But it should take the application down for sure. 

An unhandled exception in a thread is something that is different for python and c++. I need to research that more but I don't think it affects my previous statement.

**Review by wjwwood — CHANGES_REQUESTED** [review]
See my previous comment.

**Comment by fujitatomoya** [comment]
appreciate informative comments, folks! after all, i am inclined to have the default behavior to be `raise exception`.

- "when an exception occurs in a callback, it is correct to catch it and continue execution" is not really good assumption.
- user application still can manage the exception raised from executor. (means that , it can be more flexible in this way.)

(note) probably we can add doc description for each executor of `rclcpp` and `rclpy`.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17124/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11671/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17541/)

**Comment by fujitatomoya** [comment]
@wjwwood can you take a look?

**Review by ivanpauno — APPROVED** [review]
LGTM!

**Comment by ivanpauno** [comment]
> user application still can manage the exception raised from executor. (means that , it can be more flexible in this way.)

It would be nice to have an unhandled exception handler in `Executor`, as @sloretz and @wjwwood mentioned, which is not the same as that.

The part about raising an exception from the executor instead of logging and ignoring the exception seems like a safer default behavior, so LGTM!

**Comment by ivanpauno** [comment]
@fujitatomoya please check the DCO before merging

**Comment by fujitatomoya** [comment]
@ivanpauno thanks, i missed the DCO on revert commit, it's been addressed.

**Review by wjwwood — APPROVED** [review]
Lgtm, thanks for iterating

**Comment by fujitatomoya** [comment]
@Mergifyio backport humble galactic foxy

**Pull request merged**
